docs/contributing/MODEL_MANAGER.md

```diff
@@ -0,0 +1,1214 @@
+
+Another issue is that the config class hierarchy is paralleled to some
+extent by a `ModelBase` class hierarchy defined in
+`invokeai.backend.model_manager.models.base` and its subclasses. These
+are classes representing the models after they are loaded into RAM and
+include runtime information such as load status and bytes used. Some
+of the fields, including `name`, `model_type` and `base_model`, are
+shared between `ModelConfigBase` and `ModelBase`, and this is a
+potential source of confusion.
+
+** TO DO: ** The `ModelBase` code needs to be revised to reduce the
+duplication of similar classes and to support using the `key` as the
+primary model identifier.
+
+## Reading and Writing Model Configuration Records
+
+The `ModelRecordService` provides the ability to retrieve model
+configuration records from SQL or YAML databases, update them, and
+write them back.
+
+A application-wide `ModelRecordService` is created during API
+initialization and can be retrieved within an invocation from the
+`InvocationContext` object:
+
+```
+store = context.services.model_record_store
+```
+
+or from elsewhere in the code by accessing
+`ApiDependencies.invoker.services.model_record_store`.
+
+### Creating a `ModelRecordService` 
+
+To create a new `ModelRecordService` database or open an existing one,
+you can directly create either a `ModelRecordServiceSQL` or a
+`ModelRecordServiceFile` object:
+
+```
+from invokeai.app.services.model_record_service import ModelRecordServiceSQL, ModelRecordServiceFile
+
+store = ModelRecordServiceSQL.from_connection(connection, lock)
+store = ModelRecordServiceSQL.from_db_file('/path/to/sqlite_database.db')
+store = ModelRecordServiceFile.from_db_file('/path/to/database.yaml')
+```
+
+The `from_connection()` form is only available from the
+`ModelRecordServiceSQL` class, and is used to manage records in a
+previously-opened SQLITE3 database using a `sqlite3.connection` object
+and a `threading.lock` object. It is intended for the specific use
+case of storing the record information in the main InvokeAI database,
+usually `databases/invokeai.db`.
+
+The `from_db_file()` methods can be used to open new connections to
+the named database files. If the file doesn't exist, it will be
+created and initialized.
+
+As a convenience, `ModelRecordServiceBase` offers two methods,
+`from_db_file` and `open`, which will return either a SQL or File
+implementation depending on the context. The former looks at the file
+extension to determine whether to open the file as a SQL database
+(".db") or as a file database (".yaml"). If the file exists, but is
+either the wrong type or does not contain the expected schema
+metainformation, then an appropriate `AssertionError` will be raised:
+
+```
+store = ModelRecordServiceBase.from_db_file('/path/to/a/file.{yaml,db}')
+```
+
+The `ModelRecordServiceBase.open()` method is specifically designed for use in the InvokeAI
+web server and to maintain compatibility with earlier iterations of
```

Comment:
If you folks think that the `open()` API is itself ok, then I can just remove the remark about maintaining compatibility. All I meant by this is that if the user's `invokeai.yaml` config file specifies using `configs/models.yaml`, then the YAML storage implementation will be chosen so that the user has access to all their previous models. This handles legacy installations. 

New installations will default to using the sqlite backend.

Reply:
Gotcha. I'm partial to a single backend implementation, where if we see a reference to `models.yaml`, we instead migrate the app to the SQLite. But there is another thread to discuss that.